File: crates/artcraftx/src/events/functional_events/show_credential_error_modal_event.rs
use crate::events::basic_sendable_event_trait::{BasicEventStatus, BasicSendableEvent};
use crate::credentials::login_website::LoginWebsite;
use artcraft_client::enums::tauri::ux::tauri_event_name::TauriEventName;
use serde_derive::Serialize;

/// Tell the frontend to show a dismissable modal explaining a credential
/// problem: the request named no credential, the credential id didn't match
/// anything on disk, or the credential can't be used for the request.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct ShowCredentialErrorModalEvent {
  pub message: String,
  /// When the fix is to log into a website again, which one — the frontend
  /// shows a button that opens that login window.
  #[serde(skip_serializing_if = "Option::is_none")]
  pub maybe_relogin_website: Option<LoginWebsite>,
  /// The credential the re-login should refresh in place (rather than
  /// adding a second account for the same service).
  #[serde(skip_serializing_if = "Option::is_none")]
  pub maybe_credential_id: Option<String>,
}

impl BasicSendableEvent for ShowCredentialErrorModalEvent {
  const FRONTEND_EVENT_NAME: TauriEventName = TauriEventName::ShowCredentialErrorModalEvent;
  const EVENT_STATUS: BasicEventStatus = BasicEventStatus::Failure;
}

File: frontend/libs/tauri-events/src/lib/events/functional/ShowCredentialErrorModalEvent.ts
import { listen, UnlistenFn } from '@tauri-apps/api/event';
import { BasicEventWrapper } from '../../common/BasicEventWrapper';
import { useEffect } from 'react';

const EVENT_NAME : string = 'show_credential_error_modal_event';

export interface ShowCredentialErrorModalEvent {
  message: string,
  /**
   * When the fix is to log into a website again, the `LoginWebsite` value
   * (e.g. "higgsfield") to pass to `open_web_login_command`.
   */
  maybe_relogin_website?: string,
  /** The credential that re-login should refresh in place. */
  maybe_credential_id?: string,
}

export const useShowCredentialErrorModalEvent = (asyncCallback: (event: ShowCredentialErrorModalEvent) => Promise<void>) => {
  useEffect(() => {
    let isUnmounted = false;
    let unlisten: Promise<UnlistenFn>;

    const setup = async () => {
      unlisten = listen<BasicEventWrapper<ShowCredentialErrorModalEvent>>(EVENT_NAME, async (wrappedEvent) => {
        await asyncCallback(wrappedEvent.payload.data);
      });

      if (isUnmounted) {
        unlisten.then(f => f()); // Unsubscribe if unmounted early.
      }
    };

    setup();

    return () => {
      isUnmounted = true;
      unlisten.then(f => f());
    };

  }, []);
}
